File: linked_list/linked_list_max.cpp
#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* next;
        Node(int data) {
            this->data = data;
            this->next = NULL;
        }
};

class LinkedList {
    public:
        Node* head;

        LinkedList() {
            this->head = NULL;
        }

        void constructLinkedList() {
            head                   = new Node(-4);
            head->next             = new Node(-6);
            head->next->next       = new Node(-2);
            head->next->next->next = new Node(-7);
            return;
        }

        void printLinkedList() {
            if(head == NULL) {
                return;
            }

            Node* temp = head;
            while(temp) {
                cout << temp->data << " ";
                temp = temp->next;
            }
            cout << endl;
            return;
        }

        // find max in linkedlist recursively
        int findMax() {
            return findMaxUtil(head);
        }

        int findMaxUtil(Node* head) {
            if(head == NULL) {
                return 0;
            }

            if(head->next == NULL) {
                return head->data;
            }

            int curr_val = head->data;

            int max = findMaxUtil(head->next);

            if(curr_val > max) {
                max = curr_val;
            }
            return max;
        }
};


int main() {

    LinkedList linkedlist;
    linkedlist.constructLinkedList();

    linkedlist.printLinkedList();

    cout << "Max element = " << linkedlist.findMax() << endl;

    return 0;
}

File: linked_list/linked_list_max.py
"""
Find max value in linked list recursively
"""

# -*- encoding: utf-8 *-*


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Linked_list:
    def __init__(self):
        self.curr = None

    def append_node(self, curr, val):
        if curr is None:
            curr = Node(val)
            return curr
        else:
            new_node = Node(val)

            # Reach the last node
            while(curr.next is not None):
                curr = curr.next

            curr.next = new_node
            return curr

    def print_linked_list(self, curr):
        if curr is None:
            return

        while(curr is not None):
            print(curr.data, end='')
            curr = curr.next
        print()

    def find_max(self, curr):
        if curr is None:
            return

        if curr.next is None:
            return curr.data

        curr_data = curr.data

        max_data = self.find_max(curr.next)
        if curr_data > max_data:
            max_data = curr_data
        return max_data

        return None


def main():
    linked_list = Linked_list()
    linked_list.curr = linked_list.append_node(linked_list.curr, 1)
    linked_list.append_node(linked_list.curr, 2)
    linked_list.append_node(linked_list.curr, 3)
    linked_list.append_node(linked_list.curr, 4)
    linked_list.print_linked_list(linked_list.curr)
    print(linked_list.find_max(linked_list.curr))
    return


if __name__ == '__main__':
    main()

